src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala

```diff
@@ -0,0 +1,82 @@
+    */
+  def rewriteMethod(method: Method, program: Program): Method = {
+    val (expandedPres, expandablePrePredsIds) = method.pres
+      .map { expandPredicates(_, method, program) }
+      .unzip
+    val (expandedPosts, expandablePostPredsIds) = method.posts
+      .map { expandPredicates(_, method, program) }
+      .unzip
+    val expandablePrePredIds = expandablePrePredsIds.fold(Set())(_ ++ _)
+    val expandablePostPredIds = expandablePostPredsIds.fold(Set())(_ ++ _)
+    val rewrittenBody = method.body
+      .map { removeFoldUnfolds(_, expandablePrePredIds, expandablePostPredIds) }
+    method.copy(name = method.name,
+        formalArgs = method.formalArgs,
+        formalReturns = method.formalReturns,
+        pres = expandedPres,
+        posts = expandedPosts,
+        body = rewrittenBody
+      )(pos = method.pos, info = method.info, errT = method.errT)
+  }
+
+  /**
+    * Expands all predicates to their bodies in given expression.
+    * 
+    * TODO: Expand only some specified predicates, not all
+    *
+    * @param expr The expression whose predicates will be expanded
+    * @param method The method containing the expression, used to determine locally-scoped variables
+    * @param program The program containing the expression, used to expand predicates
+    * @return The expression with expanded predicates and a set of their names
+    */
+  private[this] def expandPredicates(expr: Exp, method: Method, program: Program): (Exp, Set[String]) = {
+    // Forgive me deities of functional programming for I sin
+    val expandablePredicates = mutable.Set[String]()
+    val expandedExpr = ViperStrategy.CustomContext[Set[String]]({
+      case (PredicateAccessPredicate(pred, _), args) =>
+        val body = pred.predicateBody(program, args)
+        if (body.isDefined) expandablePredicates += pred.predicateName
+        (body.getOrElse(pred), args)
+      case (quant: QuantifiedExp, args) => ???
+        // TODO: Continue traversing quant, but with context (args ++ quant.scopedDecls.map { _.name }.toSet)
+    }, method.scopedDecls.map { _.name }.toSet)
+      // .recurseFunc(???) // TODO: Do we need to call this first?
+      .execute(expr)
+    (expandedExpr, expandablePredicates.toSet)
+  }
+
+  /**
+    * Removes given predicate unfolds and folds from statement.
+    * 
+    * TODO: How to specify to Slim that the type to traverse is Stmt in general, not only Seqn?
+    *
+    * @param stmt A Seqn whose statments will be traversed
+    * @param unfoldPreds A set of the string names of the precondition predicates to not unfold
+    * @param foldPreds A set of the string names of the postcondition predicates to not fold
+    * @return The Seqn with all above unfolds and folds removed
+    */
+  private[this] def removeFoldUnfolds(stmts: Seqn, unfoldPreds: Set[String], foldPreds: Set[String]): Seqn = {
+    val noop: Stmt = Assert(TrueLit.apply()())() // TODO: Is there some other way of getting rid of a statement?
```

Comment:
I think we need to do this at the AST level, we can just traverse the list of statements manually and filter out the `Unfolds` we want to remove?